File: scraper/scraper.py
from typing import List, Dict
import requests
from bs4 import BeautifulSoup
import random
from datetime import datetime

HEADERS = {
    "User-Agent": "Mozilla/5.0",
    "Accept-Language": "en-US,en;q=0.9",
    "Accept": "text/html,application/xhtml+xml",
}

def scrape_vogue_page(url):
    results = []

    try:
        resp = requests.get(url, headers=HEADERS, timeout=15)
        resp.raise_for_status()

        soup = BeautifulSoup(resp.text, "html.parser")

        articles = soup.find_all("article")
        if not articles:
            articles = soup.find_all("div")

        for article in articles[:10]:

            title_el = article.find(["h1", "h2", "h3", "h4"])
            img_el = article.find("img")
            link_el = article.find("a", href=True)

            title = title_el.get_text(strip=True) if title_el else None

            if not title or len(title) < 5:
                continue

            img_url = ""
            if img_el:
                img_url = (
                    img_el.get("src")
                    or img_el.get("data-src")
                    or img_el.get("data-original")
                    or ""
                )

            art_url = link_el["href"] if link_el else url

            if art_url.startswith("/"):
                art_url = "https://www.vogue.com" + art_url

            product = {
                "productName": title[:120],
                "category": random.choice([
                    "Luxury Fashion", "Streetwear", "Casual Wear",
                    "Summer Collection", "Winter Collection", "Accessories"
                ]),
                "description": "Featured in Vogue – " + title[:100],
                "color": random.choice([
                    "Black", "White", "Ivory", "Beige", "Camel",
                    "Navy", "Cream", "Gold", "Red", "Blue"
                ]),
                "material": random.choice([
                    "Silk", "Cotton", "Wool", "Linen",
                    "Leather", "Cashmere", "Denim"
                ]),
                "season": random.choice([
                    "Spring/Summer", "Autumn/Winter", "All Season"
                ]),
                "imageUrl": img_url,
                "articleUrl": art_url,
                "publicationDate": datetime.now().isoformat(),
                "source": "scraped"
            }

            results.append(product)

    except Exception as e:
        print("Error:", str(e))

    return results

File: scraper/scraper.js
const https = require('https');
const http = require('http');
const { URL } = require('url');

const HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
  'Accept-Language': 'en-US,en;q=0.9',
  Accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
};

const FASHION_URLS = [
  'https://www.vogue.com/fashion',
  'https://www.vogue.com/fashion/street-style',
  'https://www.vogue.com/fashion/trends',
];

const CATEGORIES = ['Luxury Fashion', 'Streetwear', 'Casual Wear', 'Summer Collection', 'Winter Collection', 'Accessories'];
const COLORS = ['Black', 'White', 'Ivory', 'Beige', 'Camel', 'Navy', 'Cream', 'Gold', 'Red', 'Blue', 'Green', 'Multicolor'];
const MATERIALS = ['Silk', 'Cotton', 'Wool', 'Linen', 'Leather', 'Cashmere', 'Denim', 'Satin', 'Velvet', 'Metal'];
const SEASONS = ['Spring/Summer', 'Autumn/Winter', 'All Season'];

function requestPage(url) {
  return new Promise((resolve, reject) => {
    const client = url.startsWith('https') ? https : http;
    const req = client.get(url, { headers: HEADERS }, (res) => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        resolve(requestPage(new URL(res.headers.location, url).toString()));
        return;
      }

      if (res.statusCode !== 200) {
        reject(new Error(`Request failed with status ${res.statusCode}`));
        return;
      }

      let data = '';
      res.setEncoding('utf8');
      res.on('data', (chunk) => { data += chunk; });
      res.on('end', () => resolve(data));
    });

    req.on('error', reject);
    req.setTimeout(15000, () => {
      req.destroy(new Error('Request timed out'));
    });
  });
}

function stripHtml(input) {
  return input
    .replace(/<[^>]+>/g, ' ')
    .replace(/&nbsp;/g, ' ')
    .replace(/&amp;/g, '&')
    .replace(/\s+/g, ' ')
    .trim();
}

function normalizeUrl(href, baseUrl) {
  if (!href) return baseUrl;
  if (href.startsWith('http')) return href;
  if (href.startsWith('/')) return new URL(href, baseUrl).toString();
  return new URL(href, baseUrl).toString();
}

function extractArticles(html, baseUrl) {
  const results = [];
  const articlePattern = /<article\b[^>]*>([\s\S]*?)<\/article>/gi;
  const articleMatches = html.matchAll(articlePattern);

  for (const match of articleMatches) {
    const block = match[1];
    const titleMatch = block.match(/<(h1|h2|h3|h4)\b[^>]*>([\s\S]*?)<\/\1>/i);
    const linkMatch = block.match(/<a\b[^>]*href=["']([^"']+)["'][^>]*>/i);
    const imageMatch = block.match(/<img\b[^>]*src=["']([^"']+)["'][^>]*>/i);
    const title = titleMatch ? stripHtml(titleMatch[2]) : null;

    if (!title || title.length < 5) continue;

    results.push({
      title: title.slice(0, 120),
      link: normalizeUrl(linkMatch ? linkMatch[1] : '', baseUrl),
      image: imageMatch ? imageMatch[1] : '',
    });

    if (results.length >= 10) break;
  }

  return results;
}

function buildProduct(title, imageUrl, articleUrl) {
  return {
    productName: title,
    category: CATEGORIES[Math.floor(Math.random() * CATEGORIES.length)],
    description: `Featured in Vogue – ${title}.`,
    color: COLORS[Math.floor(Math.random() * COLORS.length)],
    material: MATERIALS[Math.floor(Math.random() * MATERIALS.length)],
    season: SEASONS[Math.floor(Math.random() * SEASONS.length)],
    imageUrl,
    articleUrl,
    publicationDate: new Date().toISOString(),
    source: 'scraped',
  };
}

function buildDemoData() {
  const demo = [
    { productName: 'Golden Hour Trench Coat', category: 'Luxury Fashion', color: 'Camel', material: 'Wool', season: 'Autumn/Winter' },
    { productName: 'Monochrome Midi Dress', category: 'Casual Wear', color: 'Black', material: 'Cotton', season: 'All Season' },
    { productName: 'Crystal Embellished Heels', category: 'Accessories', color: 'Silver', material: 'Metal', season: 'All Season' },
    { productName: 'Oversized Blazer Set', category: 'Streetwear', color: 'Ivory', material: 'Linen', season: 'Spring/Summer' },
  ];

  const images = [
    'https://images.unsplash.com/photo-1594938298603-c8148c4b4057?w=400&q=80',
    'https://images.unsplash.com/photo-1551537482-f2075a1d41f2?w=400&q=80',
    'https://images.unsplash.com/photo-1599643478518-a784e5dc4c8f?w=400&q=80',
    'https://images.unsplash.com/photo-1572804013309-59a88b7e92f1?w=400&q=80',
  ];

  return demo.map((item, index) => ({
    ...item,
    description: `As featured in Vogue – ${item.productName} is a must-have piece this season.`,
    imageUrl: images[index % images.length],
    articleUrl: 'https://www.vogue.com/fashion',
    publicationDate: new Date().toISOString(),
    source: 'scraped',
  }));
}

async function scrapeVogue() {
  const allProducts = [];

  for (const url of FASHION_URLS) {
    try {
      console.error(`Scraping: ${url}`);
      const html = await requestPage(url);
      const articles = extractArticles(html, url);
      for (const article of articles) {
        allProducts.push(buildProduct(article.title, article.image, article.link));
      }
      await new Promise((resolve) => setTimeout(resolve, 1200 + Math.random() * 1300));
    } catch (error) {
      console.error(`Failed to scrape ${url}: ${error.message}`);
    }
  }

  const seen = new Set();
  const unique = [];
  for (const product of allProducts) {
    if (!seen.has(product.productName)) {
      seen.add(product.productName);
      unique.push(product);
    }
  }

  if (unique.length === 0) {
    console.error('Live scraping returned no results – using demo data.');
    return buildDemoData();
  }

  return unique;
}

scrapeVogue()
  .then((products) => {
    process.stdout.write(JSON.stringify(products));
  })
  .catch((error) => {
    console.error(error);
    process.exit(1);
  });
